plugins/ua_network_pubsub_ethernet.c

```diff
@@ -0,0 +1,422 @@
+
+    /* set destination MAC address */
+    if (ether_aton_r(channelDataEth->destinationMac,
+            (struct ether_addr*)&ethHdr->ether_dhost) == (struct ether_addr*) 0) {
+        UA_LOG_ERROR(UA_Log_Stdout, UA_LOGCATEGORY_SERVER,
+            "PubSub connection send failed. Cannot set destination MAC address.");
+        UA_free(bufSend);
+        return UA_STATUSCODE_BADINTERNALERROR;
+    }
+
+    /* set ethertype */
+    /* either VLAN or UADP */
+    ptrCur = bufSend + sizeof(*ethHdr);
+    if (channelDataEth->vid == 0) {
+        ethHdr->ether_type = htons(ETHERTYPE_UADP);
+        lenBuf -= 4;  /* no VLAN tag */
+    } else {
+        ethHdr->ether_type = htons(ETHERTYPE_VLAN);
+        /* set VLAN ID */
+        UA_UInt16 vlanTag;
+        vlanTag = (UA_UInt16) (channelDataEth->vid + (channelDataEth->prio << 13));
+        *((UA_UInt16 *) ptrCur) = htons(vlanTag);
+        ptrCur += sizeof(UA_UInt16);
+        /* set UADP */
+        *((UA_UInt16 *) ptrCur) = htons(ETHERTYPE_UADP);
+        ptrCur += sizeof(UA_UInt16);
+    }
+
+    /* copy payload of ethernet message */
+    memcpy(ptrCur, buf->data, buf->length);
+
+    ssize_t rc;
+    rc = UA_send(channel->sockfd, bufSend, lenBuf, 0);
+    if (rc  < 0) {
+        UA_LOG_ERROR(UA_Log_Stdout, UA_LOGCATEGORY_SERVER,
+            "PubSub connection send failed. Send message failed.");
+        UA_free(bufSend);
+        return UA_STATUSCODE_BADINTERNALERROR;
+    }
+    UA_free(bufSend);
+
+    return UA_STATUSCODE_GOOD;
+}
+
+/**
+ * Receive messages.
+ *
+ * @param timeout in usec -> not used
+ * @return
+ */
+static UA_StatusCode
+UA_PubSubChannelEthernet_receive(UA_PubSubChannel *channel, UA_ByteString *message, UA_ExtensionObject *transportSettings, UA_UInt32 timeout) {
+    static unsigned char buffer[2048];
+    ssize_t dataLen;
+
+    (void) transportSettings;
+    (void) timeout;
+
+    dataLen = UA_recv(channel->sockfd, buffer, sizeof(buffer), MSG_DONTWAIT);
+    if (dataLen < 0) {
+        if ((errno == EAGAIN) || (errno == EWOULDBLOCK)) {
+            return UA_STATUSCODE_GOODNODATA;
+        }
+        UA_LOG_ERROR(UA_Log_Stdout, UA_LOGCATEGORY_SERVER,
+            "PubSub connection receive failed. Receive message failed.");
+        return UA_STATUSCODE_BADINTERNALERROR;
+    } else if (dataLen == 0) {
+        return UA_STATUSCODE_GOODNODATA;
+    }
+
+    /* handle VLAN tag */
+    struct ether_header *ethHdr;
+    ssize_t ethHdrLen;
+    UA_UInt16 etherType;
+
+    ethHdr = (struct ether_header*) buffer;
+    ethHdrLen = sizeof(*ethHdr);
+    etherType = ntohs(ethHdr->ether_type);
+    if (etherType == ETHERTYPE_VLAN) {
+        ethHdrLen += 4;
+        etherType = *((UA_UInt16 *) (ethHdr + ethHdrLen + 2));
+    }
+    if (etherType != ETHERTYPE_UADP) {
+        UA_LOG_ERROR(UA_Log_Stdout, UA_LOGCATEGORY_SERVER,
+            "PubSub connection receive failed. Not UADP format.");
+        return UA_STATUSCODE_BADINTERNALERROR;
```

Comment:
With ETH_P_ALL any messages could be received here, without being an error. Statuscode should be something like SKIP or GOOD_NO_MSG.